# AutoML for Beginners with H2O
### Titanic - Machine Learning from Disaster Challenge | Kaggle

This is a Jupyter Notebook. When you execute code within the notebook, the results appear beneath the code. To execute a code chunk, place your cursor on the cell and press Shift+Enter

This is the famous **"Titanic - Machine Learning from Disaster"** competition from [Kaggle](https://www.kaggle.com/c/titanic/overview). If you are new to Machine Learning this is one of the best, first challenges for you to dive into Machine Learning.
The challenge is to use machine learning to create a model that predicts which passengers survived the Titanic shipwreck.
We are going to do this by using [H2O AutoML](https://docs.h2o.ai/h2o/latest-stable/h2o-docs/automl.html).

[H2O](https://www.h2o.ai) is a fully open-source, distributed in-memory machine learning platform with linear scalability. H2O supports the most widely used statistical & machine learning algorithms including gradient boosted machines, generalized linear models, deep learning and more. H2O also has an industry-leading AutoML functionality that automatically runs through all the algorithms and their hyperparameters to produce a leaderboard of the best models.

H2O’s AutoML can be used for automating the machine learning workflow, which includes automatic training and tuning of many models within a user-specified time limit.

H2O has made it easy for non-experts to experiment with machine learning through AutoMl. But, there is still a fair bit of knowledge and background in data science that is required to produce high-performing machine learning models.
It is also useful for advanced users. H2O helps them by freeing up their time to focus on other aspects of the data science pipeline tasks such as data-preprocessing, feature engineering and model deployment.

### RMS Titanic

"RMS Titanic was a British passenger liner operated by the White Star Line that sank in the North Atlantic Ocean on 15 April 1912, after striking an iceberg during her maiden voyage from Southampton to New York City. Of the estimated 2,224 passengers and crew aboard, more than 1,500 died, making the sinking at the time one of the deadliest of a single ship and the deadliest peacetime sinking of a superliner or cruise ship to date." -  [Wikipedia](https://en.wikipedia.org/wiki/Titanic)

## Prerequisites

We are going to do this challenge using H2O in Python. The prerequisites for this is as follows,
<ol>
    <li>
        <p>Operating Systems</p>
        <ul>
            <li>Windows 7 or later</li>
            <li>OS X 10.9 or later</li>
            <li>Ubuntu 12.04 or later</li>
            <li>RHEL/CentOS 6 or later</li>
        </ul>
    </li>
    <li>
        <p>Language</p>
        <ul>
            <li>Python 2.7.x, 3.5.x, 3.6.x</li>
        </ul>
    </li>
</ol>

You need Java installed and JAVA_HOME environment variable set in your system to run H2O.

Download and install Java([Amazon Correto](https://aws.amazon.com/corretto/)) from [here](https://docs.aws.amazon.com/corretto/latest/corretto-8-ug/downloads-list.html).

The installer will automatically set the JAVA_HOME environment variable for you.

<em>
Amazon Corretto is a no-cost, multiplatform, production-ready distribution of the Open Java Development Kit (OpenJDK). Corretto comes with long-term support that will include performance enhancements and security fixes.
</em>

You can download Python 3.6 from here: [Windows(exe)](https://www.python.org/ftp/python/3.6.6/python-3.6.6-amd64.exe)/[Linux(source)](https://www.python.org/ftp/python/3.6.6/Python-3.6.6.tgz)

## Installing H20

First you need to install the dependencies (prepending with `sudo` if needed):

In [1]:
#Remove ! from each line if you are running the commands from CLI
!pip install requests
!pip install tabulate
!pip install "colorama>=0.3.8"
!pip install future

You are using pip version 18.1, however version 21.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


You are using pip version 18.1, however version 21.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


You are using pip version 18.1, however version 21.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


You are using pip version 18.1, however version 21.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


The following command removes any existing installation of H2O module from Python.

In [2]:
#Remove ! from each line if you are running the commands from CLI
!pip uninstall h2o

^C


Next, use pip to install this version of the H2O Python module.

In [3]:
#Remove ! from each line if you are running the commands from CLI
!pip install http://h2o-release.s3.amazonaws.com/h2o/rel-zipf/1/Python/h2o-3.32.1.1-py2.py3-none-any.whl

You are using pip version 18.1, however version 21.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


To install this package with conda, uncomment the second line and run:

In [ ]:
#Remove ! from each line if you are running the commands from CLI
#!conda install -c h2oai h2o

## Start H2O

Import the H2O Python module and `H2OAutoML` class and initialize a local H2O cluster.

In [1]:
import h2o
from h2o.automl import H2OAutoML
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  ---------------------------------------------------------
H2O_cluster_uptime:         13 mins 05 secs
H2O_cluster_timezone:       Asia/Kolkata
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.32.1.1
H2O_cluster_version_age:    10 days
H2O_cluster_name:           H2O_from_python_nidhi_fxqxxl
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    2.434 Gb
H2O_cluster_total_cores:    8
H2O_cluster_allowed_cores:  8
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.6.8 final
--------------------------  ---------------------------------------------------------

## Load Data

Download and copy the dataset to a subfolder named "data" in the same directory as your Python script/Jupiter notebook.
The dataset contains two files **train.csv** and **test.csv**.

We will use the training set(train.csv) to build our machine learning models. For the training set, we provide the outcome (also known as the “ground truth”) for each passenger. Our model will be based on “features” like passengers’ gender and class.

The test set is to see how well our model performs on unseen data. For the test set, we do not provide the ground truth for each passenger. It is our job to predict these outcomes. For each passenger in the test set, we will use the model that we trained to predict whether or not they survived the sinking of the Titanic.

### Data Dictionary

|Variable|Defenition|Key|
|:--------|:----------|:----|
|survival|Survival|0 = No, 1 = Yes|
|pclass|Ticket class|1 = 1st, 2 = 2nd, 3 = 3rd|
|sex|Sex| |
|Age|Age in years| |
|sibsp|# of siblings / spouses aboard the Titanic| |
|parch|# of parents / children aboard the Titanic| |
|ticket|Ticket number| |
|fare|Passenger fare| |
|cabin|Cabin number| |
|embarked|Port of Embarkation|C = Cherbourg, Q = Queenstown, S = Southampton|

Source: [Kaggle](https://www.kaggle.com/c/titanic/data)

Now we will load the data into H20.

In [2]:
# Use local data file
data_path = "./data/train.csv"

# Load data into H2O
df = h2o.import_file(data_path)

Parse progress: |█████████████████████████████████████████████████████████| 100%


### Handling Missing Data and Adding New Features

Next, we have to deal with missing values. If you look at the missing row in the above output, there are 687 missing records in Cabin, 177 in Age, and 2 in Embarked columns.

A cabin number looks like ‘A123’ and the letter refers to the deck. Therefore we’re going to extract these and create a new feature called Deck.

In [3]:
df["Deck"] = df["Cabin"].strsplit("([0-9]*)")[0]
df.describe()

Rows:891
Cols:13




PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked Deck 
 
 
 type int int int string enum real int int int real enum enum enum 
 mins 1.0 0.0 1.0 NaN 0.42 0.0 0.0 693.0 0.0 
 mean 446.0 0.3838383838383838 2.3086419753086447 NaN 29.69911764705884 0.5230078563411893 0.3815937149270483 260318.5491679275 32.20420796857465 
 maxs 891.0 1.0 3.0 NaN 80.0 8.0 6.0 3101298.0 512.3292 
 sigma 257.3538420152301 0.4865924542648575 0.8360712409770491 NaN 14.526497332334035 1.1027434322934315 0.8060572211299488 471609.26868834975 49.69342859718089 
 zeros 0 549 0 0 0 608 678 0 15 
 missing 0 0 0 0 0 177 0 0 230 0 687 2 687 
 0 1.0 0.0 3.0 Braund, Mr. Owen Harris male 22.0 1.0 0.0 nan 7.25 S 
 1 2.0 1.0 1.0 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38.0 1.0 0.0 nan 71.2833 C85 C C 
 2 3.0 1.0 3.0 Heikkinen, Miss. Laina female 26.0 0.0 0.0 nan 7.925 S 
 3 4.0 1.0 1.0 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35.0 1.0 0.0 113803.0 53.1 C123 S C 
 4 5.0 0.0 3.0 Allen, Mr. William Henry male 35.0 0.0 0.0 373450.0 8.05 S 
 5 6.0 0.0 3.0 Moran, Mr. James male nan 0.0 0.0 330877.0 8.4583 Q 
 6 7.0 0.0 1.0 McCarthy, Mr. Timothy J male 54.0 0.0 0.0 17463.0 51.8625 E46 S E 
 7 8.0 0.0 3.0 Palsson, Master. Gosta Leonard male 2.0 3.0 1.0 349909.0 21.075 S 
 8 9.0 1.0 3.0 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27.0 0.0 2.0 347742.0 11.1333 S 
 9 10.0 1.0 2.0 Nasser, Mrs. Nicholas (Adele Achem) female 14.0 1.0 0.0 237736.0 30.0708 C

Next, we will calculate the missing age of passengers. I am going to replace the missing values with the mean value along with a group by column Pclass.

In [5]:
df.impute("Age", method="mean", by=["Pclass"])
df.describe()

Rows:891
Cols:13




PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked Deck 
 
 
 type int int int string enum real int int int real enum enum enum 
 mins 1.0 0.0 1.0 NaN 0.42 0.0 0.0 693.0 0.0 
 mean 446.0 0.3838383838383838 2.3086419753086447 NaN 29.292874790272105 0.5230078563411893 0.3815937149270483 260318.5491679275 32.20420796857465 
 maxs 891.0 1.0 3.0 NaN 80.0 8.0 6.0 3101298.0 512.3292 
 sigma 257.3538420152301 0.4865924542648575 0.8360712409770491 NaN 13.210527334401528 1.1027434322934315 0.8060572211299488 471609.26868834975 49.69342859718089 
 zeros 0 549 0 0 0 608 678 0 15 
 missing 0 0 0 0 0 0 0 0 230 0 687 2 687 
 0 1.0 0.0 3.0 Braund, Mr. Owen Harris male 22.0 1.0 0.0 nan 7.25 S 
 1 2.0 1.0 1.0 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38.0 1.0 0.0 nan 71.2833 C85 C C 
 2 3.0 1.0 3.0 Heikkinen, Miss. Laina female 26.0 0.0 0.0 nan 7.925 S 
 3 4.0 1.0 1.0 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35.0 1.0 0.0 113803.0 53.1 C123 S C 
 4 5.0 0.0 3.0 Allen, Mr. William Henry male 35.0 0.0 0.0 373450.0 8.05 S 
 5 6.0 0.0 3.0 Moran, Mr. James male 25.14061971830986 0.0 0.0 330877.0 8.4583 Q 
 6 7.0 0.0 1.0 McCarthy, Mr. Timothy J male 54.0 0.0 0.0 17463.0 51.8625 E46 S E 
 7 8.0 0.0 3.0 Palsson, Master. Gosta Leonard male 2.0 3.0 1.0 349909.0 21.075 S 
 8 9.0 1.0 3.0 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27.0 0.0 2.0 347742.0 11.1333 S 
 9 10.0 1.0 2.0 Nasser, Mrs. Nicholas (Adele Achem) female 14.0 1.0 0.0 237736.0 30.0708 C

Since the Embarked column has only 2 missing values, we will just replace the missing rows with the most common value.

In [6]:
df.impute("Embarked", method="mode")
df.describe()

Rows:891
Cols:13




PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked Deck 
 
 
 type int int int string enum real int int int real enum enum enum 
 mins 1.0 0.0 1.0 NaN 0.42 0.0 0.0 693.0 0.0 
 mean 446.0 0.3838383838383838 2.3086419753086447 NaN 29.292874790272105 0.5230078563411893 0.3815937149270483 260318.5491679275 32.20420796857465 
 maxs 891.0 1.0 3.0 NaN 80.0 8.0 6.0 3101298.0 512.3292 
 sigma 257.3538420152301 0.4865924542648575 0.8360712409770491 NaN 13.210527334401528 1.1027434322934315 0.8060572211299488 471609.26868834975 49.69342859718089 
 zeros 0 549 0 0 0 608 678 0 15 
 missing 0 0 0 0 0 0 0 0 230 0 687 0 687 
 0 1.0 0.0 3.0 Braund, Mr. Owen Harris male 22.0 1.0 0.0 nan 7.25 S 
 1 2.0 1.0 1.0 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38.0 1.0 0.0 nan 71.2833 C85 C C 
 2 3.0 1.0 3.0 Heikkinen, Miss. Laina female 26.0 0.0 0.0 nan 7.925 S 
 3 4.0 1.0 1.0 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35.0 1.0 0.0 113803.0 53.1 C123 S C 
 4 5.0 0.0 3.0 Allen, Mr. William Henry male 35.0 0.0 0.0 373450.0 8.05 S 
 5 6.0 0.0 3.0 Moran, Mr. James male 25.14061971830986 0.0 0.0 330877.0 8.4583 Q 
 6 7.0 0.0 1.0 McCarthy, Mr. Timothy J male 54.0 0.0 0.0 17463.0 51.8625 E46 S E 
 7 8.0 0.0 3.0 Palsson, Master. Gosta Leonard male 2.0 3.0 1.0 349909.0 21.075 S 
 8 9.0 1.0 3.0 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27.0 0.0 2.0 347742.0 11.1333 S 
 9 10.0 1.0 2.0 Nasser, Mrs. Nicholas (Adele Achem) female 14.0 1.0 0.0 237736.0 30.0708 C

Next, I am going to add some more new features by combining values from the existing features.

#### AgePclass

In [8]:
df["AgePclass"] = df["Age"]*df["Pclass"]

#### NoOfFamilyMembers

In [9]:
df["NoOfFamilyMembers"] = (df["Parch"]*df["SibSp"])

#### IsAlone

In [10]:
df["IsAlone"] = df["NoOfFamilyMembers"]/df["NoOfFamilyMembers"]
df[df["IsAlone"].isna(), "IsAlone"] = 0

#### FarePerPerson

In [11]:
df["FarePerPerson"] = df["Fare"]/(df["NoOfFamilyMembers"]+1)

Now I am going to update the missing values of the Deck feature with a new value. First, let's see what all alphabets are currently used in this field.

In [12]:
 df["Deck"].table()

Deck Count 
 
 
 A 15 
 B 47 
 C 59 
 D 33 
 T 1 
 E 32 
 F 13 
 G 4

We'll update empty Deck values with "Z".

In [13]:
df["Deck"] = df["Deck"].ascharacter()
df[df["Deck"].isna(), "Deck"] = "Z"

For classification, the Survived column should be encoded as categorical (aka. "factor" or "enum"). Let's take a look.

In [17]:
df.describe()

Rows:891
Cols:17




PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked Deck AgePclass NoOfFamilyMembers IsAlone FarePerPerson 
 
 
 type int enum enum string enum real int int int real enum enum enum real int enum real 
 mins 1.0 NaN 0.42 0.0 0.0 693.0 0.0 0.92 0.0 0.0 
 mean 446.0 NaN 29.292874790272105 0.5230078563411893 0.3815937149270483 260318.5491679275 32.20420796857465 63.17118286436423 0.5679012345679009 26.545188724601847 
 maxs 891.0 NaN 80.0 8.0 6.0 3101298.0 512.3292 222.0 16.0 512.3292 
 sigma 257.3538420152301 NaN 13.210527334401528 1.1027434322934315 0.8060572211299488 471609.26868834975 49.69342859718089 31.498476922565953 1.9792865515522966 44.63917637309987 
 zeros 0 0 0 608 678 0 15 0 749 15 
 missing 0 0 0 0 0 0 0 0 230 0 687 0 0 0 0 0 0 
 0 1.0 0 3 Braund, Mr. Owen Harris male 22.0 1.0 0.0 nan 7.25 S Z 66.0 0.0 0 7.25 
 1 2.0 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38.0 1.0 0.0 nan 71.2833 C85 C C 38.0 0.0 0 71.2833 
 2 3.0 1 3 Heikkinen, Miss. Laina female 26.0 0.0 0.0 nan 7.925 S Z 78.0 0.0 0 7.925 
 3 4.0 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35.0 1.0 0.0 113803.0 53.1 C123 S C 35.0 0.0 0 53.1 
 4 5.0 0 3 Allen, Mr. William Henry male 35.0 0.0 0.0 373450.0 8.05 S Z 105.0 0.0 0 8.05 
 5 6.0 0 3 Moran, Mr. James male 25.14061971830986 0.0 0.0 330877.0 8.4583 Q Z 75.42185915492958 0.0 0 8.4583 
 6 7.0 0 1 McCarthy, Mr. Timothy J male 54.0 0.0 0.0 17463.0 51.8625 E46 S E 54.0 0.0 0 51.8625 
 7 8.0 0 3 Palsson, Master. Gosta Leonard male 2.0 3.0 1.0 349909.0 21.075 S Z 6.0 3.0 1 5.26875 
 8 9.0 1 3 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27.0 0.0 2.0 347742.0 11.1333 S Z 81.0 0.0 0 11.1333 
 9 10.0 1 2 Nasser, Mrs. Nicholas (Adele Achem) female 14.0 1.0 0.0 237736.0 30.0708 C Z 28.0 0.0 0 30.0708

By default that cell got processed as an integer. Pclass, Deck and IsAlone should also be enum as it represents the socio-economic status of passengers. We need to convert these into an enum. We can do that by just executing the following line,

In [18]:
df["Survived"]= df["Survived"].asfactor()
df["Pclass"]= df["Pclass"].asfactor()
df["IsAlone"]= df["IsAlone"].asfactor()
df["Deck"] = df["Deck"].asfactor()

Now let's see what happend.

In [19]:
df.describe()

Rows:891
Cols:17




PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked Deck AgePclass NoOfFamilyMembers IsAlone FarePerPerson 
 
 
 type int enum enum string enum real int int int real enum enum enum real int enum real 
 mins 1.0 NaN 0.42 0.0 0.0 693.0 0.0 0.92 0.0 0.0 
 mean 446.0 NaN 29.292874790272105 0.5230078563411893 0.3815937149270483 260318.5491679275 32.20420796857465 63.17118286436423 0.5679012345679009 26.545188724601847 
 maxs 891.0 NaN 80.0 8.0 6.0 3101298.0 512.3292 222.0 16.0 512.3292 
 sigma 257.3538420152301 NaN 13.210527334401528 1.1027434322934315 0.8060572211299488 471609.26868834975 49.69342859718089 31.498476922565953 1.9792865515522966 44.63917637309987 
 zeros 0 0 0 608 678 0 15 0 749 15 
 missing 0 0 0 0 0 0 0 0 230 0 687 0 0 0 0 0 0 
 0 1.0 0 3 Braund, Mr. Owen Harris male 22.0 1.0 0.0 nan 7.25 S Z 66.0 0.0 0 7.25 
 1 2.0 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38.0 1.0 0.0 nan 71.2833 C85 C C 38.0 0.0 0 71.2833 
 2 3.0 1 3 Heikkinen, Miss. Laina female 26.0 0.0 0.0 nan 7.925 S Z 78.0 0.0 0 7.925 
 3 4.0 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35.0 1.0 0.0 113803.0 53.1 C123 S C 35.0 0.0 0 53.1 
 4 5.0 0 3 Allen, Mr. William Henry male 35.0 0.0 0.0 373450.0 8.05 S Z 105.0 0.0 0 8.05 
 5 6.0 0 3 Moran, Mr. James male 25.14061971830986 0.0 0.0 330877.0 8.4583 Q Z 75.42185915492958 0.0 0 8.4583 
 6 7.0 0 1 McCarthy, Mr. Timothy J male 54.0 0.0 0.0 17463.0 51.8625 E46 S E 54.0 0.0 0 51.8625 
 7 8.0 0 3 Palsson, Master. Gosta Leonard male 2.0 3.0 1.0 349909.0 21.075 S Z 6.0 3.0 1 5.26875 
 8 9.0 1 3 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27.0 0.0 2.0 347742.0 11.1333 S Z 81.0 0.0 0 11.1333 
 9 10.0 1 2 Nasser, Mrs. Nicholas (Adele Achem) female 14.0 1.0 0.0 237736.0 30.0708 C Z 28.0 0.0 0 30.0708

As you can see the data type of Survived and Pclass columns are now enum. Let's leave all the categorical features to be auto handled by H20.

### Generate x and Y values

Now, let's identify the response & predictor columns by saving them as x and y. The PassengerId, Name and Ticket columns are unique identifiers so we'll remove those from the set of our predictors along with the Cabin feature.

In [20]:
y = "Survived"
x = df.columns
x.remove(y)
x.remove("PassengerId")
x.remove("Name")
x.remove("Ticket")
x.remove("Cabin")

## Run AutoML

Run AutoML, stopping after 15 models. The max_models argument specifies the number of individual (or "base") models and does not include the two ensemble models that are trained at the end. We set a seed value for reproducibility. H2O Deep Learning models are not reproducible by default for performance reasons, so for reproducibility, we are adding "DeepLearning" to exclude_algos parameter. The exploitation_ratio parameter is for specifying the budget ratio (between 0 and 1) dedicated to the exploitation (vs exploration) phase. By default, the exploitation phase is disabled (exploitation_ratio=0) as this is still experimental; to activate it, it is recommended to try a ratio around 0.1. Learn more about exploitation and exploration from [here](https://www.manifold.ai/exploration-vs-exploitation-in-reinforcement-learning).

In [21]:
aml = H2OAutoML(max_models = 15, seed = 1, nfolds=15, exclude_algos = ["DeepLearning"], exploitation_ratio = 0.15)
aml.train(x = x, y = y, training_frame = df)

AutoML progress: |█
17:34:22.242: AutoML: XGBoost is not available; skipping it.

███████████████████████████████████████████████████████| 100%


## Leaderboard

Next, we will view the AutoML Leaderboard. Since we did not specify a leaderboard_frame in the H2OAutoML.train() method for scoring and ranking the models, the AutoML leaderboard uses cross-validation metrics to rank the models.

A default performance metric for each machine learning task (binary classification, multiclass classification, regression) is specified internally and the leaderboard will be sorted by that metric. In the case of binary classification, the default ranking metric is Area Under the ROC Curve (AUC). In the future, the user will be able to specify any of the H2O metrics so that different metrics can be used to generate rankings on the leaderboard.

The leader model is stored at aml.leader and the leaderboard is stored at aml.leaderboard.

In [22]:
lb = aml.leaderboard

Now we will view a snapshot of the top models. Here we should see the two Stacked Ensembles at or near the top of the leaderboard. Stacked Ensembles can almost always outperform a single model.

In [23]:
lb.head()

model_id auc logloss aucpr mean_per_class_error rmse mse 
 
 
 GBM_2_AutoML_20210405_173422 0.876117 0.406868 0.857066 0.178219 0.354391 0.125593 
 StackedEnsemble_AllModels_AutoML_20210405_173422 0.875771 0.40771 0.855449 0.174624 0.354071 0.125366 
 StackedEnsemble_BestOfFamily_AutoML_20210405_173422 0.875337 0.405848 0.852315 0.178434 0.353587 0.125024 
 GBM_grid__1_AutoML_20210405_173422_model_3 0.871681 0.420932 0.847572 0.188165 0.362799 0.131623 
 GBM_4_AutoML_20210405_173422 0.871231 0.416213 0.852867 0.172874 0.356901 0.127378 
 GBM_3_AutoML_20210405_173422 0.871044 0.411429 0.852986 0.173929 0.356137 0.126834 
 DRF_1_AutoML_20210405_173422 0.870954 0.471875 0.835348 0.174839 0.363062 0.131814 
 GBM_grid__1_AutoML_20210405_173422_model_6 0.869516 0.460762 0.843318 0.191808 0.373499 0.139502 
 GBM_grid__1_AutoML_20210405_173422_model_2 0.868682 0.412404 0.85504 0.187997 0.357784 0.12801 
 GBM_1_AutoML_20210405_173422 0.868378 0.421288 0.848769 0.171556 0.360635 0.130058

To view the entire leaderboard, specify the `rows` argument of the `head()` method as the total number of rows:

In [24]:
lb.head(rows=lb.nrows)

model_id auc logloss aucpr mean_per_class_error rmse mse 
 
 
 GBM_2_AutoML_20210405_173422 0.876117 0.406868 0.857066 0.178219 0.354391 0.125593 
 StackedEnsemble_AllModels_AutoML_20210405_173422 0.875771 0.40771 0.855449 0.174624 0.354071 0.125366 
 StackedEnsemble_BestOfFamily_AutoML_20210405_173422 0.875337 0.405848 0.852315 0.178434 0.353587 0.125024 
 GBM_grid__1_AutoML_20210405_173422_model_3 0.871681 0.420932 0.847572 0.188165 0.362799 0.131623 
 GBM_4_AutoML_20210405_173422 0.871231 0.416213 0.852867 0.172874 0.356901 0.127378 
 GBM_3_AutoML_20210405_173422 0.871044 0.411429 0.852986 0.173929 0.356137 0.126834 
 DRF_1_AutoML_20210405_173422 0.870954 0.471875 0.835348 0.174839 0.363062 0.131814 
 GBM_grid__1_AutoML_20210405_173422_model_6 0.869516 0.460762 0.843318 0.191808 0.373499 0.139502 
 GBM_grid__1_AutoML_20210405_173422_model_2 0.868682 0.412404 0.85504 0.187997 0.357784 0.12801 
 GBM_1_AutoML_20210405_173422 0.868378 0.421288 0.848769 0.171556 0.360635 0.130058 
 GBM_grid__1_AutoML_20210405_173422_model_4 0.867398 0.432804 0.843982 0.195187 0.368007 0.135429 
 GBM_grid__1_AutoML_20210405_173422_model_5 0.867151 0.418324 0.849888 0.165732 0.354813 0.125892 
 XRT_1_AutoML_20210405_173422 0.864637 0.438892 0.834533 0.18306 0.372122 0.138475 
 GBM_grid__1_AutoML_20210405_173422_model_1 0.863867 0.475881 0.828301 0.186871 0.374516 0.140262 
 GBM_5_AutoML_20210405_173422 0.859987 0.435374 0.835221 0.199094 0.370209 0.137055 
 GLM_1_AutoML_20210405_173422 0.852574 0.44767 0.829721 0.203744 0.37653 0.141775 
 GBM_grid__1_AutoML_20210405_173422_model_7 0.850401 0.463847 0.81912 0.22203 0.385372 0.148511

As we can see the leader in this example is [GBM](https://towardsdatascience.com/understanding-gradient-boosting-machines-9be756fe76ab) which stands for Gradient Boosting Machines with an AUC value of 0.876117. 

AUC ranges in value from 0 to 1. A model whose predictions are 100% wrong has an AUC of 0.0; one whose predictions are 100% correct has an AUC of 1.0. Learn more about AUC from [here](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc).

You can google and see what all are other values and how they affect the accuracy of our model.

## Save Leader Model

There are two ways to save the leader model -- binary format and MOJO format. If you're taking your leader model to production, then I'd suggest the MOJO format since it's optimized for production use.

In [25]:
h2o.save_model(aml.leader, path = "./out/titanic_survivability_model_bin")

'F:\\Projects\\nidhinradh.me\\Code\\Titanic\\out\\titanic_survivability_model_bin\\GBM_2_AutoML_20210405_173422'

In [26]:
import os
mojo_out_folder = "./out/titanic_survivability_model_mojo/"
mojo_file_name = "mojo_out.zip"
try:
    os.remove(mojo_out_folder + mojo_file_name)
except OSError:
    pass
os.makedirs(mojo_out_folder, exist_ok=True)
from zipfile import ZipFile
with ZipFile(mojo_out_folder + mojo_file_name, 'w') as file:
    pass
aml.leader.download_mojo(path = mojo_out_folder + mojo_file_name)

'F:\\Projects\\nidhinradh.me\\Code\\Titanic\\out\\titanic_survivability_model_mojo\\mojo_out.zip'

## Prediction

In [ ]:
In this step we will load the test.csv file to H2O and predict the survivability of each passenger.

In [33]:
# Use local data file
test_data_path = "./data/test.csv"

# Load data into H2O
test_df = h2o.import_file(test_data_path)

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [34]:
test_df.describe()

Rows:418
Cols:11




PassengerId Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked 
 
 
 type int int string enum real int int int real enum enum 
 mins 892.0 1.0 NaN 0.17 0.0 0.0 680.0 0.0 
 mean 1100.5 2.2655502392344493 NaN 30.272590361445786 0.44736842105263147 0.39234449760765555 223850.98986486494 35.627188489208656 
 maxs 1309.0 3.0 NaN 76.0 8.0 9.0 3101298.0 512.3292 
 sigma 120.81045760473994 0.8418375519640504 NaN 14.181209235624422 0.8967595611217135 0.9814288785371694 369523.7764694362 55.907576179973844 
 zeros 0 0 0 0 283 324 0 2 
 missing 0 0 0 0 86 0 0 122 1 327 0 
 0 892.0 3.0 Kelly, Mr. James male 34.5 0.0 0.0 330911.0 7.8292 Q 
 1 893.0 3.0 Wilkes, Mrs. James (Ellen Needs) female 47.0 1.0 0.0 363272.0 7.0 S 
 2 894.0 2.0 Myles, Mr. Thomas Francis male 62.0 0.0 0.0 240276.0 9.6875 Q 
 3 895.0 3.0 Wirz, Mr. Albert male 27.0 0.0 0.0 315154.0 8.6625 S 
 4 896.0 3.0 Hirvonen, Mrs. Alexander (Helga E Lindqvist) female 22.0 1.0 1.0 3101298.0 12.2875 S 
 5 897.0 3.0 Svensson, Mr. Johan Cervin male 14.0 0.0 0.0 7538.0 9.225 S 
 6 898.0 3.0 Connolly, Miss. Kate female 30.0 0.0 0.0 330972.0 7.6292 Q 
 7 899.0 2.0 Caldwell, Mr. Albert Francis male 26.0 1.0 1.0 248738.0 29.0 S 
 8 900.0 3.0 Abrahim, Mrs. Joseph (Sophie Halaut Easu) female 18.0 0.0 0.0 2657.0 7.2292 C 
 9 901.0 3.0 Davies, Mr. John Samuel male 21.0 2.0 0.0 nan 24.15 S

We'll add the newly created features and convert Pclass as we did for the training dataset. You can also see that a row is missing the value for the fare. I am going to update the missing value with mean along with a group by column Pclass.

In [37]:
test_df["Deck"] = test_df["Cabin"].strsplit("([0-9]*)")[0]
test_df["AgePclass"] = test_df["Age"]*test_df["Pclass"]
test_df["NoOfFamilyMembers"] = (test_df["Parch"]*test_df["SibSp"])
test_df["IsAlone"] = test_df["NoOfFamilyMembers"]/test_df["NoOfFamilyMembers"]
test_df[test_df["IsAlone"].isna(), "IsAlone"] = 0
test_df["FarePerPerson"] = test_df["Fare"]/(test_df["NoOfFamilyMembers"]+1)

test_df["Pclass"]= test_df["Pclass"].asfactor()
test_df.impute("Fare", method="mean", by=["Pclass"]) 
test_df.describe()

Rows:418
Cols:16




PassengerId Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked Deck AgePclass NoOfFamilyMembers IsAlone FarePerPerson 
 
 
 type int enum string enum real int int int real enum enum enum real int int real 
 mins 892.0 NaN 0.17 0.0 0.0 680.0 0.0 0.51 0.0 0.0 0.0 
 mean 1100.5 NaN 30.272590361445786 0.44736842105263147 0.39234449760765555 223850.98986486494 35.571763822679856 59.03337349397588 0.4449760765550237 0.15311004784688995 28.825602374336853 
 maxs 1309.0 NaN 76.0 8.0 9.0 3101298.0 512.3292 181.5 16.0 1.0 512.3292 
 sigma 120.81045760473994 NaN 14.181209235624422 0.8967595611217135 0.9814288785371694 369523.7764694362 55.85199678933088 29.04879625409189 1.6687524955824409 0.3605250532203249 46.70526282058175 
 zeros 0 0 0 283 324 0 2 0 354 354 2 
 missing 0 0 0 0 86 0 0 122 0 327 0 327 86 0 0 1 
 0 892.0 3 Kelly, Mr. James male 34.5 0.0 0.0 330911.0 7.8292 Q 103.5 0.0 0.0 7.8292 
 1 893.0 3 Wilkes, Mrs. James (Ellen Needs) female 47.0 1.0 0.0 363272.0 7.0 S 141.0 0.0 0.0 7.0 
 2 894.0 2 Myles, Mr. Thomas Francis male 62.0 0.0 0.0 240276.0 9.6875 Q 124.0 0.0 0.0 9.6875 
 3 895.0 3 Wirz, Mr. Albert male 27.0 0.0 0.0 315154.0 8.6625 S 81.0 0.0 0.0 8.6625 
 4 896.0 3 Hirvonen, Mrs. Alexander (Helga E Lindqvist) female 22.0 1.0 1.0 3101298.0 12.2875 S 66.0 1.0 1.0 6.14375 
 5 897.0 3 Svensson, Mr. Johan Cervin male 14.0 0.0 0.0 7538.0 9.225 S 42.0 0.0 0.0 9.225 
 6 898.0 3 Connolly, Miss. Kate female 30.0 0.0 0.0 330972.0 7.6292 Q 90.0 0.0 0.0 7.6292 
 7 899.0 2 Caldwell, Mr. Albert Francis male 26.0 1.0 1.0 248738.0 29.0 S 52.0 1.0 1.0 14.5 
 8 900.0 3 Abrahim, Mrs. Joseph (Sophie Halaut Easu) female 18.0 0.0 0.0 2657.0 7.2292 C 54.0 0.0 0.0 7.2292 
 9 901.0 3 Davies, Mr. John Samuel male 21.0 2.0 0.0 nan 24.15 S 63.0 0.0 0.0 24.15

We will now remove the unique identifier columns from the dataset.

In [38]:
test_df = test_df[:, ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked", "Deck", "AgePclass", "NoOfFamilyMembers", "IsAlone", "FarePerPerson"]]

Using the predict() function with AutoML generates predictions on the leader model from the run. The order of the rows in the results is the same as the order in which the data was loaded, even if some rows fail (for example, due to missing values or unseen factor levels).

In [39]:
# To generate predictions on a test set, you can make predictions
# directly on the `"H2OAutoML"` object or on the leader model
# object directly
preds = aml.predict(test_df)

# or:
# preds = aml.leader.predict(test)
preds

gbm prediction progress: |████████████████████████████████████████████████| 100%


predict p0 p1 
 
 
 0 0.958668 0.0413324 
 0 0.804113 0.195887 
 0 0.891065 0.108935 
 0 0.640217 0.359783 
 0 0.743113 0.256887 
 0 0.838345 0.161655 
 1 0.435593 0.564407 
 0 0.967373 0.0326268 
 1 0.289667 0.710333 
 0 0.91355 0.0864501

p0 is the probability (between 0 and 1) that class 0 is chosen and p1 is the probability (between 0 and 1) that class 1 is chosen.

You can use this Dataframe for further analysis.

## Conclusion

Hooray! We have covered almost all the portions that you need to know for getting started with AutoML. AutoML is the process of automating the time consuming, iterative tasks of machine learning model development. It allows data scientists, analysts, and developers of any experience level to build ML models with high scale, efficiency, and productivity all while sustaining model quality.
We have started with the installation of H2O, then we loaded the data to H20, checked missing data, updated those missing values and created new features. Then we trained the dataset using H2O AutoML with 15 different models and found out that GBM is a good candidate for handling this specific dataset. We also learned how to save this model and to use it for making predictions later.
There are many things you can still do to improve this piece of code. You can do more feature engineering, standardize the numerical values, find the correlation between the dependant and independent variables, etc.